File: tests/router_test.py
# -*- coding: utf-8 -*-

import sys, os
sys.path.append(os.path.dirname(__file__))
sys.path.append(os.path.dirname(os.path.dirname(__file__)))

from oktest import ok, test, subject, situation, at_end

from minikeight import (
    Router,
    NaiveLinearRouter, PrefixLinearRouter, FixedLinearRouter,
    NaiveRegexpRouter, SmartRegexpRouter, NestedRegexpRouter,
    StateMachineRouter,
)
from mock_handler import HomeAPI, BooksAPI, BookCommentsAPI, OrdersAPI, mapping


class MockRouter(Router):

    def find(self, req_path):
        d = {
            "/":
                (HomeAPI, {"GET": HomeAPI.do_home}, []),
            "/api/v1/books.json":
                (BooksAPI, {"GET": BooksAPI.do_index, "POST": BooksAPI.do_create}, []),
            "/api/v1/books/123.json":
                (BooksAPI, {"GET": BooksAPI.do_show, "PUT": BooksAPI.do_update, "DELETE": BooksAPI.do_delete}, [123]),
            "/api/v1/orders/":
                (OrdersAPI, {"GET": OrdersAPI.do_index, "POST": OrdersAPI.do_create}, []),
            "/api/v1/orders/123.html":
                (OrdersAPI, {"GET": OrdersAPI.do_show, "PUT": OrdersAPI.do_update, "DELETE": OrdersAPI.do_delete}, [123]),
            "/api/v1/orders/123.json":
                (OrdersAPI, {"GET": OrdersAPI.do_show, "PUT": OrdersAPI.do_update, "DELETE": OrdersAPI.do_delete}, [123]),
        }
        return d.get(req_path)


class Router_TestCase(object):

    def provide_router(self):
        return MockRouter()

    with subject('#lookup()'):

        @test("returns tuple when found.")
        def _(self, router):
            t = router.lookup("GET", "/api/v1/books.json")
            ok (t) == (BooksAPI, BooksAPI.do_index, [])
            t = router.lookup("POST", "/api/v1/books.json")
            ok (t) == (BooksAPI, BooksAPI.do_create, [])
            t = router.lookup("GET", "/api/v1/books/123.json")
            ok (t) == (BooksAPI, BooksAPI.do_show, [123])
            t = router.lookup("PUT", "/api/v1/books/123.json")
            ok (t) == (BooksAPI, BooksAPI.do_update, [123])
            t = router.lookup("DELETE", "/api/v1/books/123.json")
            ok (t) == (BooksAPI, BooksAPI.do_delete, [123])
            #
            t = router.lookup("GET", "/api/v1/orders/")
            ok (t) == (OrdersAPI, OrdersAPI.do_index, [])
            t = router.lookup("POST", "/api/v1/orders/")
            ok (t) == (OrdersAPI, OrdersAPI.do_create, [])
            t = router.lookup("GET", "/api/v1/orders/123.html")
            ok (t) == (OrdersAPI, OrdersAPI.do_show, [123])
            t = router.lookup("PUT", "/api/v1/orders/123.html")
            ok (t) == (OrdersAPI, OrdersAPI.do_update, [123])
            t = router.lookup("DELETE", "/api/v1/orders/123.html")
            ok (t) == (OrdersAPI, OrdersAPI.do_delete, [123])


class Router_TestBase(object):

    ROUTER_CLASS = None
    TUPLE_TYPE   = tuple

    def provide_router(self):
        return self.ROUTER_CLASS(mapping)

    def provide_tupletype(self):
        return self.TUPLE_TYPE

    def _test_when_found(self, router):
        tupletype = self.TUPLE_TYPE
        c = HomeAPI
        t = router.find('/')
        methods = {"GET": c.do_home}
        ok (t) == (c, methods, [])
        ##
        c = BooksAPI
        t = router.find('/api/v1/books.json')
        methods = {"GET": c.do_index, "POST": c.do_create}
        ok (t) == (c, methods, [])
        #
        t = router.find('/api/v1/books/123.json')
        methods = {"GET": c.do_show, "PUT": c.do_update, "DELETE": c.do_delete}
        ok (t) == (c, methods, [123])
        ##
        c = OrdersAPI
        t = router.find('/api/v1/orders/')
        methods = {"GET": c.do_index, "POST": c.do_create}
        ok (t) == (c, methods, [])
        #
        t = router.find('/api/v1/orders/123.json')
        methods = {"GET": c.do_show, "PUT": c.do_update, "DELETE": c.do_delete}
        ok (t) == (c, methods, [123])
        ##
        c = BookCommentsAPI
        t = router.find('/api/v1/books/123/comments')
        methods = {"GET": c.do_index, "POST": c.do_create}
        ok (t) == (c, methods, [123])
        #
        t = router.find('/api/v1/books/123/comments/abcd')
        methods = {"GET": c.do_show, "PUT": c.do_update, "DELETE": c.do_delete}
        ok (t) == (c, methods, [123, 'abcd'])

    def _test_when_not_found(self, router):
        ok (router.find('')) == None
        #
        ok (router.find('/api/v1/books/')) == None
        ok (router.find('/api/v1/books/abc.json')) == None
        ok (router.find('/api/v1/books/123.html')) == None
        #
        ok (router.find('/api/v1/orders')) == None
        ok (router.find('/api/v1/orders/abc.html')) == None
        ok (router.find('/api/v1/orders/123.html/')) == None
        #
        ok (router.find('/api/v1/books/')) == None
        ok (router.find('/api/v1/books/123/comments/abcd.json')) == None
        ok (router.find('/api/v1/books/123/comments/abcd/')) == None

    def _test_suffix_pattern(self, router):
        tupletype = self.TUPLE_TYPE
        c = OrdersAPI
        methods = {"GET": c.do_show, "PUT": c.do_update, "DELETE": c.do_delete}
        #
        t = router.find('/api/v1/orders/123.json')
        ok (t) == (c, methods, [123])
        #
        t = router.find('/api/v1/orders/123.html')
        ok (t) == (c, methods, [123])
        #
        t = router.find('/api/v1/orders/123')
        ok (t) == (c, methods, [123])

    with subject("#find()"):

        @test("returns tuple when found.")
        def _(self, router):
            self._test_when_found(router)

        @test("returns None when not found.")
        def _(self, router):
            self._test_when_not_found(router)

        @test("supports '.*' as suffix pattern.")
        def _(self, router):
            self._test_suffix_pattern(router)


class NaiveLinearRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = NaiveLinearRouter


class PrefixLinearRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = PrefixLinearRouter


class FixedLinearRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = FixedLinearRouter


class NaiveRegexpRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = NaiveRegexpRouter


class SmartRegexpRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = SmartRegexpRouter


class NestedRegexpRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = NestedRegexpRouter


class StateMachineRouter_TestCase(Router_TestBase):
    ROUTER_CLASS = StateMachineRouter
    TUPLE_TYPE = staticmethod(lambda xs: [ (int(x) if x.isdigit() else x) for x in xs ])


if __name__ == '__main__':
    import oktest
    oktest.main()

File: tests/ruby/router_test.rb
# coding: utf-8

require 'oktest'
require_relative '../../minikeight'


class HelloHandler < K8::RequestHandler
  mapping "/"          , :GET=>:do_index, :POST=>:do_create
  mapping "/new"       , :GET=>:do_new
  mapping "/{id}.json" , :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete
  mapping "/{id}/edit" , :GET=>:do_edit
  def do_index()   ; {action: "index"}         ; end
  def do_create()  ; {action: "create"}        ; end
  def do_new()     ; {action: "new"}           ; end
  def do_show(id)  ; {action: "show", id: id}  ; end
  def do_update(id); {action: "update", id: id}; end
  def do_delete(id); {action: "delete", id: id}; end
  def do_edit(id)  ; {action: "edit", id: id}  ; end
end

class CommentHandler < K8::RequestHandler
  mapping "/"          , :GET=>:do_index, :POST=>:do_create
  mapping "/new"       , :GET=>:do_new
  mapping "/{id}.json" , :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete
  mapping "/{id}/edit" , :GET=>:do_edit
  def do_index(x_id)   ; {action: "index"}         ; end
  def do_create(x_id)  ; {action: "create"}        ; end
  def do_new(x_id)     ; {action: "new"}           ; end
  def do_show(x_id, comment_id)  ; {action: "show", id: id}  ; end
  def do_update(x_id, comment_id); {action: "update", id: id}; end
  def do_delete(x_id, comment_id); {action: "delete", id: id}; end
  def do_edit(x_id, comment_id)  ; {action: "edit", id: id}  ; end
end

class AnyHandler < K8::RequestHandler
  mapping "" ,:GET=>:do_nothing
  def do_nothing(*args)
    return {args=>args}
  end
end

mapping1 = {
  "/api/v1" => {
    "/hello" => HelloHandler,
    "/hello/{x_id}/comments" => CommentHandler,
    "/helloworld" => HelloHandler,
  },
  ##
  "/repos/{owner}/{repo}" => {
    "/git/blobs" => AnyHandler,
    "/pulls/{pull_number}/reviews/{review_id}/comments" => AnyHandler,
    "/generate" => AnyHandler,
  },
}
mapping2 = {
  "/api/v2" => {
    "/hello" => {
      "" => HelloHandler,
      "/{x_id}/comments" => CommentHandler,
      "world" => HelloHandler,
    }
  },
  ##
  "/repos/{owner}" => {
    "/{repo}" => {
      "/git/blobs" => AnyHandler,
      "/pulls/{pull_number}" => {
        "/reviews/{review_id}/comments" => AnyHandler,
      },
      "/generate" => AnyHandler,
    },
  },
}

router_classes = [
  K8::NaiveLinearRouter,
  K8::PrefixLinearRouter,
  K8::FixedLinearRouter,
  K8::HashedLinearRouter,
  K8::NaiveRegexpRouter,
  K8::SmartRegexpRouter,
  K8::NestedRegexpRouter,
  K8::OptimizedRegexpRouter,
  K8::SlicedRegexpRouter,
  K8::HashedRegexpRouter,
  K8::ListBasedTrieRouter,
  K8::DictBasedTrieRouter,
  K8::StateMachineRouter,
  K8::NestedDictRouter,
]


Oktest.scope do

  maketestcases = proc do |klass, mapping, ver|

    topic klass do

      router = klass.new(mapping)

      topic '#find()' do

        spec "/api/#{ver}/hello/" do
          ret = router.find("/api/#{ver}/hello/")
          methods = {:GET=>:do_index, :POST=>:do_create}
          ok {ret} == [HelloHandler, methods, []]
        end

        spec "/api/#{ver}/hello/new" do
          ret = router.find("/api/#{ver}/hello/new")
          methods = {:GET=>:do_new}
          ok {ret} == [HelloHandler, methods, []]
        end

        spec "/api/#{ver}/hello/123.json" do
          ret = router.find("/api/#{ver}/hello/123.json")
          methods = {:GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete}
          ok {ret} == [HelloHandler, methods, [123]]
        end

        spec "/api/#{ver}/hello/123/edit" do
          ret = router.find("/api/#{ver}/hello/123/edit")
          methods = {:GET=>:do_edit}
          ok {ret} == [HelloHandler, methods, [123]]
        end

        spec "/api/#{ver}/hello/123/comments/" do
          ret = router.find("/api/#{ver}/hello/123/comments/")
          methods = {:GET=>:do_index, :POST=>:do_create}
          ok {ret} == [CommentHandler, methods, [123]]
        end

        spec "/api/#{ver}/hello/123/comments/{id}.json" do
          ret = router.find("/api/#{ver}/hello/123/comments/888.json")
          methods = {:GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete}
          ok {ret} == [CommentHandler, methods, [123, 888]]
        end

        spec "/api/#{ver}/helloworld/" do
          ret = router.find("/api/#{ver}/helloworld/")
          methods = {:GET=>:do_index, :POST=>:do_create}
          ok {ret} == [HelloHandler, methods, []]
        end

        spec "/api/#{ver}/helloworld/123.json" do
          ret = router.find("/api/#{ver}/helloworld/123.json")
          methods = {:GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete}
          ok {ret} == [HelloHandler, methods, [123]]
        end

        ##

        spec "/repos/owner1/repo2/pulls/333/reviews/444/comments" do
          ret = router.find("/repos/owner1/repo2/pulls/333/reviews/444/comments")
          methods = {:GET=>:do_nothing}
          ok {ret} == [AnyHandler, methods, ["owner1", "repo2", "333", 444]]
        end

      end

    end

  end

  router_classes.each do |router_class|
    {"v1" => mapping1, "v2" => mapping2}.each do |ver, mapping|
      maketestcases.call(router_class, mapping, ver)
    end
  end

end
